### DISCUSSION TITLE
Way to build symbolic linear state-space models from transfer functions

### DISCUSSION THREAD

**Discussion by billtubbs** [casadi/casadi#3754] [Q&A]
I'd like to know what's the best way to build the symbolic matrices of a linear state space model from transfer functions, either in CasADi or using some external library.

I don't see a TransferFunction or LinearStateSpace class in CasADi (has anyone written such tools?) but my goal would be to do something like this:

```python
import casadi as *

K = SX.sym('K')
T1 = SX.sym('T1')

G = TransferFunction(K, 1 + T1*s, s)  # doesn't exist
sys = StateSpace(G)  # doesn't exist
print(sys.A, sys.B, sys.C, sys.D)
```
Expected output:
```
(-1/T1) 1 K/T1 00
```

I noticed that there is a new [control module to Sympy](https://github.com/sympy/sympy/issues/19352).  So I can build a symbolic state-space model in Sympy as follows:

```python
import sympy
from sympy.physics.control.lti import TransferFunction, StateSpace

s, K, T1 = sympy.symbols("s, K, T1")

G = TransferFunction(K, 1 + T1*s, s)
sys = G.rewrite(StateSpace)
print(sys)
```
Output:
```
StateSpace(Matrix([[-1/T1]]), Matrix([[1]]), Matrix([[K]]), Matrix([[0]]))
```

But then how would I convert these to (possibly sparse) CasADi matrices?  Has anyone written a Sympy -> Casadi expression converter?

Or are there any external control libraries that accept CasADi symbolic variables?

Any advice or comments welcome.

**Comment by billtubbs** [comment]
I just found this: [sympy2casadi](https://gist.github.com/jgillis/80bb594a6c8fcf55891d1d88b12b68b8#file-sympy2casadi)

It seems to work with a few modifications:

```python
import sympy
from sympy.physics.control.lti import TransferFunction, StateSpace
import casadi as cas

def sympy2casadi(sympy_expr, sympy_vars, casadi_vars, sparsify=True):
    """Convert Sympy expression to CasADi symbolic expression.

    Warning: This function uses Python's exec function, and thus should not
    be used on unsanitized input.
    """

    mapping = {
        'ImmutableDenseMatrix': cas.blockcat,
        'MutableDenseMatrix': cas.blockcat,
        'Abs': cas.fabs
    }
    f = sympy.lambdify(sympy_vars, sympy_expr, modules=[mapping, cas])

    result = f(*casadi_vars)
    if sparsify:
        return cas.sparsify(result)
    else:
        return result


s, K, T1 = sympy.symbols("s, K, T1")
G = TransferFunction(K, 1 + T1*s, s)
sys = G.rewrite(StateSpace)

K_cas = cas.SX.sym('K')
T1_cas = cas.SX.sym('T1')
var_map = {K: K_cas, T1: T1_cas}

A, B, C, D = [
    sympy2casadi(item, var_map.keys(), var_map.values()) 
    for item in [sys.state_matrix, sys.input_matrix, sys.output_matrix , sys.feedforward_matrix]
]
print(A, B, C, D)
```

Output:
```
(-1/T1) 1 K 00
```

**Reply by namannimmo10** [reply]
Glad you found a solution!

**Comment by billtubbs** [comment]
FYI, I think there is a bug in Sympy.  The result above is incorrect.  See https://github.com/sympy/sympy/issues/26827.
